feat: send 'add_to_cart' analytics event

heitorado opened the pull request:
## What's the purpose of this pull request?
This PR uses the methods defined on the store sdk related to sending/receiving analytics events, to send the 'add_to_cart' event. This event is triggered every time a product is added to the shopping cart, and then received by an analytics provider that stores the data. The data that is sent is a merge between the fields required by SAE analytics and by GA4, so it can be used by both.

## How it works? 
The `ProductDetails.ts` and `ProductSummary.ts` passes the required analytics + cart attributes to the `useBuyButton` hook. When the button is pressed, the attributes are sent as an analytics event, which then is listened by the `AnalyticsHandler` and inserted in the `window.dataLayer` array to be consumed by the Analytics service (GA4 or SAE).
After sending the event with the cart attributes plus some extra attributes required by the analytics, the object needs to be destructured and built again so it only has the fields accepted to make the 'add to cart' mutation on the `store-api` side. This is why a new object is build before sending the product to the `addItem` method.

## How to test it?
`yarn test`. New tests added on `cart.test.js`, covering:
- The analytics event is added when clicking on the buy button at the PLP;
- The analytics event is added when clicking on the buy button at the PDP;
- The `add_to_cart` analytics event is sent with at least the minimum required attributes. 
To test it manually, add something into the shopping cart. Then, open the console and check if the `window.dataLayer` array has the correct analytics data for the added item.

## References
https://www.notion.so/vtexhandbook/vtex-addToCart-8125848d75cb4a34adbfa57199fc4b66
https://www.dumky.net/posts/analytics-and-tag-testing-with-cypress/
https://docs.cypress.io/guides/references/assertions#TDD-Assertions
https://docs.cypress.io/guides/references/assertions#BDD-Assertions
https://www.chaijs.com/api/bdd/#method_property

A big thanks to @icazevedo for the help with GA4 and the great documentation of the issues/tasks; also to @tlgimenes for all the help to understand the codebase.

Commit:
feat: send 'add_to_cart' analytics event

add logic to send 'add_to_cart' analytics event as a callback when buy button is pressed

Commit:
feat: create AnalyticsProvider component

Commit:
feat: add AnalyticsProvider to gatsby ssr and browser

Commit:
test: add cypress tests for add_to_cart analytics event

icazevedo commented on src/sdk/analytics/index.tsx:
```diff
@@ -0,0 +1,17 @@
+import { useAnalyticsEvent } from '@vtex/store-sdk'
+import type { FC } from 'react'
+
+export const AnalyticsProvider: FC = ({ children }) => {
```
I think it'd be better if we renamed this component to something like `AnalyticsHandler`. It doesn't provide a context, which is when we usually name things as providers in React. I think leaving it this way may confuse people :)

heitorado replied:
That makes a lot of sense. Done at https://github.com/vtex-sites/base.store/pull/53/commits/d1f33648e138fbb01faeed277229284a22e3405a

heitorado replied:
@tlgimenes here is where the `AnalyticsProvider` was renamed to `AnalyticsHandler`

icazevedo commented on src/sdk/cart/useBuyButton.ts:
```diff
@@ -1,12 +1,34 @@
+}
+
+interface VTEXAddToCartEvent extends AddToCartEvent {
+  data: AddToCartEvent['data'] & {
```
I figured out that on store-sdk we export an interface called `AddToCartData` that can replace this call. I think it'd be cleaner :)

```suggestion
  data: AddToCartData & {
```

heitorado replied:
Great. I believe this solves it? https://github.com/vtex-sites/base.store/pull/53/commits/e84887be8e1813e5866d198221006e7a930f3990

icazevedo commented on src/sdk/cart/useBuyButton.ts:
```diff
@@ -16,10 +38,37 @@ export const useBuyButton = (item: Omit<CartItem, 'id'> | null) => {
+        type: 'add_to_cart',
+        data: {
+          currency: code,
+          value: item.price * item.quantity,
```
I think it's worth leaving a `TODO` comment here saying that in the future we may explore more robust ways of calculating the value. Maybe there is some gift item or some discount depending on the quantity. I don't think we should worry about that now, but it's valid to document the possibility.

heitorado replied:
Fair enough. Done at https://github.com/vtex-sites/base.store/pull/53/commits/d46f50c9cdb1aeee730a2742744a2f8742c89b98

icazevedo reviewed (changes requested):
Great work 👏

heitorado commented on src/sdk/analytics/index.tsx:
```diff
@@ -0,0 +1,17 @@
+import type { FC } from 'react'
+
+export const AnalyticsProvider: FC = ({ children }) => {
+  useEffect(() => {
+    window.dataLayer = window.dataLayer || []
+  })
```
Thanks for the heads up! Still exploring out the world of react hooks 👍🏽

heitorado committed:
fix: add second parameter on useEffect hook

Co-authored-by: Ícaro Azevedo <icaro.azevedo@vtex.com.br>

Commit:
fix: return children directly on AnalyticsProvider

instead of returning it wrapped into a react fragment

Commit:
fix: rename AnalyticsProvider to AnalyticsHandler

Commit:
docs: add todo comment
to explore other ways of calculating the product's value in the future

heitorado commented on src/sdk/analytics/index.tsx:
```diff
@@ -0,0 +1,17 @@
+    window.dataLayer.push(event)
+  })
+
+  return <> {children} </>
```
You're right. Removed the `FC` return type and added the `PropsWithChildren<unknown>` type to the argument. Then,   TypeScript's type inference does the trick. Done at https://github.com/vtex-sites/base.store/pull/53/commits/31c27fb1fc2d6a15505afadb3bf1564cdf665d92

Commit:
refactor: change interface call to a cleaner one

instead of AddToCartEvent['data'] it is cleaner to use AddToCartData

icazevedo reviewed (approved):
LGTM, but I miss a deeper test to check if the properties of the event are really there and if they are in the correct format. Is it possible to do it?

tlgimenes commented on gatsby-ssr.js:
```diff
@@ -6,23 +6,26 @@ import ErrorBoundary from './src/sdk/error/ErrorBoundary'
-        </SessionProvider>
-      </UIProvider>
-    </TestProvider>
+    <AnalyticsHandler>
```
Why not use the standard and call it `AnalyticsProvider` ?

heitorado replied:
I agree it would 'match' with the other providers being used, but on a previous comment made by @icazevedo he pointed out that given the component implementation, the name "Provider" would not suit it well. I've tagged you in the conversation thread.

tlgimenes commented on src/components/product/ProductSummary/ProductSummary.tsx:
```diff
@@ -77,6 +80,10 @@ export const fragment = graphql`
     slug: linkText
     id: productId
     productName
+    brand
```
I think these types have changed since last time. Please rebase.

tlgimenes commented on src/sdk/analytics/index.tsx:
```diff
@@ -0,0 +1,17 @@
+
+export const AnalyticsHandler = ({ children }: PropsWithChildren<unknown>) => {
+  useEffect(() => {
+    window.dataLayer = window.dataLayer || []
```
This makes a race condition possible if an analytics event is run before this effect. To elimitate this problem please move the dataLayer creation to the global scope. Also, make sure to make it work durring SSR
```suggestion
if (typeof window !== 'undefined') {
   window.dataLayer = window.dataLayer || []
}

export const AnalyticsHandler = ({ children }: PropsWithChildren<unknown>) => {
  useAnalyticsEvent((event) => {
    window.dataLayer.push(event)
  })
```

Commit:
fix: adjust event to work with store-api

some graphql queries changed. Removed 'categoryTree' from queries since it is no longer available. Also, changed the query format for 'brand'.

heitorado committed:
fix: move dataLayer creation to global scope

Co-authored-by: Tiago Gimenes <tlgimenes@gmail.com>

Commit:
fix: correct AnalyticsHandler merge

Commit:
test: add cypress tests to check relevant event fields

emersonlaurentino commented:
You need make a rebase on your branch

Commit:
fix: change fields passed to useBuyButton

Commit:
fix: regenerate generated graphql files

Commit:
fix: remove brand from add_to_cart event

igorbrasileiro reviewed (approved):
Great job, @heitorado! 👏 🚀

Commit:
feat: Add brand and item name to analytics event

This was achieved by destructuring and mounting the item object again with only the properties allowed by store-api. The useBuyButton sends extra fields to analytics, and those fields need to be removed before calling the 'addItem' method that triggers the mutation that adds the item to the store-sdk cart.

Commit:
fix: remove sku_reference_id from analytics event item
